### ISSUE TITLE
Can't compile 3.11.x with pyenv: "undefined symbol: sqlite3_deserialize"

### ISSUE THREAD

**Issue by pcroland** [pyenv/pyenv#2625]
- Platform information: Ubuntu 20.04.5 LTS (WSL)
- OS architecture: x64
- pyenv version: 2.3.13
- Python version: trying to compile 3.11.2
- C Compiler information: gcc 9.4.0

When I do `pyenv install 3.11.2` (or 3.11.1) it fails with the following message:
```
pyenv install 3.11.2
pyenv: /home/roland/.pyenv/versions/3.11.2 already exists
continue with installation? (y/N) y
Installing Python-3.11.2...
Traceback (most recent call last):
  File "<string>", line 1, in <module>
  File "/home/roland/.pyenv/versions/3.11.2/lib/python3.11/sqlite3/__init__.py", line 57, in <module>
    from sqlite3.dbapi2 import *
  File "/home/roland/.pyenv/versions/3.11.2/lib/python3.11/sqlite3/dbapi2.py", line 27, in <module>
    from _sqlite3 import *
ModuleNotFoundError: No module named '_sqlite3'
WARNING: The Python sqlite3 extension was not compiled. Missing the SQLite3 lib?
Installed Python-3.11.2 to /home/roland/.pyenv/versions/3.11.2
```
I have everything installed from the `Common build problems` wiki, like `libsqlite3-dev`.

output of `env PYENV_DEBUG=1 pyenv install -v 3.11.2 2>&1 | tee trace.log`: [trace.log](https://up.ovpn.to/file/x89YAXdPms4FKnyO1ii3dwHeR2AZgCdCCmpHOKeU/trace.log)

**Comment by native-api** [comment]
There's a problem linking with `libcrypt` and `libsqlite3` at your machine.

```
*** WARNING: renaming "_crypt" since importing it failed: libcrypt.so.2: cannot open shared object file: No such file or directory
*** WARNING: renaming "_sqlite3" since importing it failed: /tmp/python-build.20230219131449.1221/Python-3.11.2/build/lib.linux-x86_64-3.11/_sqlite3.cpython-311-x86_64-linux-gnu.so: undefined symbol: sqlite3_deserialize
```

Here, the extensions link with them like this:

```
$ ldd /home/vmuser/.pyenv/versions/3.11.2/lib/python3.11/lib-dynload/_crypt.cpython-311-x86_64-linux-gnu.so
	linux-vdso.so.1 (0x00007ffe2b9cf000)
	libcrypt.so.1 => /lib/x86_64-linux-gnu/libcrypt.so.1 (0x00007fd140d3b000)
	libc.so.6 => /lib/x86_64-linux-gnu/libc.so.6 (0x00007fd140b13000)
	/lib64/ld-linux-x86-64.so.2 (0x00007fd140d8b000)
$ ldd /home/vmuser/.pyenv/versions/3.11.2/lib/python3.11/lib-dynload/_sqlite3.cpython-311-x86_64-linux-gnu.so
	linux-vdso.so.1 (0x00007ffd358c5000)
	libsqlite3.so.0 => /lib/x86_64-linux-gnu/libsqlite3.so.0 (0x00007fcfba4a9000)
	libc.so.6 => /lib/x86_64-linux-gnu/libc.so.6 (0x00007fcfba281000)
	libm.so.6 => /lib/x86_64-linux-gnu/libm.so.6 (0x00007fcfba19a000)
	/lib64/ld-linux-x86-64.so.2 (0x00007fcfba627000)
```

Note that:

* My OS is Ubuntu 22.04, newer that yours, and yet it still uses `libcrypt.so.1`, not `libcrypt.so.2`
* [`sqlite3_deserialize` is an optional feature](https://www.sqlite.org/compile.html#omit_deserialize) -- but CPython's configure does check for `sqlite3_serialize` which is always present together with it, and it did found it.

**Do you by any chance have alternative version of `libcrypt` and `libsqlite3` installed -- e.g. to `/usr/local` -- which override the system libraries?**

**Comment by pcroland** [comment]
`which sqlite3` returned `/home/linuxbrew/.linuxbrew/bin/sqlite`. I removed it, now it points to `/usr/bin/sqlite3` but the compilation error remains. [trace.log](https://up.ovpn.to/file/KAboi08ghyG9khZkucVnKCLNv7go5S9QtS4VEVXs/trace.log)

**Comment by ushuz** [comment]
I encountered a similar problem installing Python 3.11.2 on WSL (Ubuntu 20.04.5), but with `_ssl`. However, I can install 3.10.10 without issue after being unable to install 3.11.2.

**Comment by promlife** [comment]
I have the same problem

**Comment by ajgringo619** [comment]
I can't get this to fail on either Arch or Debian. Since both are using newer versions of **GCC** (12 and 10, respectively), could that be the problem? Or maybe this is specific to **WSL**, which I don't have access to at the moment.

**Comment by Xav-Pe** [comment]
Same problem, with 3.11.1. It tells me some modules are missing : _bz2, _curses, _ctypes, etc.


    Platform information: Ubuntu 20.04.1 LTS (WSL)
    OS architecture: x64
    pyenv version: 2.3.13



```console
❯ pyenv install 3.11.1
Downloading Python-3.11.1.tar.xz...
-> https://www.python.org/ftp/python/3.11.1/Python-3.11.1.tar.xz
Installing Python-3.11.1...
Traceback (most recent call last):
  File "<string>", line 1, in <module>
  File "/home/xp/.pyenv/versions/3.11.1/lib/python3.11/bz2.py", line 17, in <module>
    from _bz2 import BZ2Compressor, BZ2Decompressor
ModuleNotFoundError: No module named '_bz2'
WARNING: The Python bz2 extension was not compiled. Missing the bzip2 lib?
Traceback (most recent call last):
  File "<string>", line 1, in <module>
  File "/home/xp/.pyenv/versions/3.11.1/lib/python3.11/curses/__init__.py", line 13, in <module>
    from _curses import *
ModuleNotFoundError: No module named '_curses'
WARNING: The Python curses extension was not compiled. Missing the ncurses lib?
Traceback (most recent call last):
  File "<string>", line 1, in <module>
  File "/home/xp/.pyenv/versions/3.11.1/lib/python3.11/ctypes/__init__.py", line 8, in <module>
    from _ctypes import Union, Structure, Array
ModuleNotFoundError: No module named '_ctypes'
WARNING: The Python ctypes extension was not compiled. Missing the libffi lib?
Traceback (most recent call last):
  File "<string>", line 1, in <module>
ModuleNotFoundError: No module named 'readline'
WARNING: The Python readline extension was not compiled. Missing the GNU readline lib?
Traceback (most recent call last):
  File "<string>", line 1, in <module>
  File "/home/xp/.pyenv/versions/3.11.1/lib/python3.11/ssl.py", line 100, in <module>
    import _ssl             # if we can't import it, let the error propagate
    ^^^^^^^^^^^
ModuleNotFoundError: No module named '_ssl'
ERROR: The Python ssl extension was not compiled. Missing the OpenSSL lib?

Please consult to the Wiki page to fix the problem.
https://github.com/pyenv/pyenv/wiki/Common-build-problems
```

**Comment by pcroland** [comment]
> Rerun the installation with `-k` (so that it retains the build directory). In the build directory (which would now be under `$PYENV_ROOT/sources`), find the corresponding extension modules (they are dynamic libraries). Run `LD_DEBUG=1 ldd <module>` for each one. It should tell you why it cannot load.
I don't know which file to call `LD_DEBUG` on, could you reopen the issue?

**Comment by native-api** [comment]
> *** WARNING: renaming "_crypt" since importing it failed: libcrypt.so.2: cannot open shared object file: No such file or directory
> *** WARNING: renaming "_sqlite3" since importing it failed: /tmp/python-build.20230219131449.1221/Python-3.11.2/build/lib.linux-x86_64-3.11/_sqlite3.cpython-311-x86_64-linux-gnu.so: undefined symbol: sqlite3_deserialize

They should be dynamic libraries (with a name changed in some way) that have `_crypt` and `_sqlite3` in their names.

**Comment by pcroland** [comment]
```sh
find . -iname "*_sqlite*"
./3.11.3/Python-3.11.3/Lib/test/test_sqlite3
./3.11.3/Python-3.11.3/Modules/_sqlite
./3.11.3/Python-3.11.3/PCbuild/_sqlite3.vcxproj
./3.11.3/Python-3.11.3/PCbuild/_sqlite3.vcxproj.filters
./3.11.3/Python-3.11.3/build/lib.linux-x86_64-3.11/_sqlite3.cpython-311-x86_64-linux-gnu_failed.so
./3.11.3/Python-3.11.3/build/temp.linux-x86_64-3.11/home/roland/.pyenv/sources/3.11.3/Python-3.11.3/Modules/_sqlite```

**Comment by native-api** [comment]
`_sqlite3.cpython-311-x86_64-linux-gnu_failed.so` is a clear winner

**Comment by pcroland** [comment]
```sh
LD_DEBUG=1 ldd _crypt.cpython-311-x86_64-linux-gnu_failed.so _sqlite3.cpython-311-x86_64-linux-gnu_failed.so
warning: debug option `1' unknown; try LD_DEBUG=help
_crypt.cpython-311-x86_64-linux-gnu_failed.so:
warning: debug option `1' unknown; try LD_DEBUG=help
warning: debug option `1' unknown; try LD_DEBUG=help
        linux-vdso.so.1 (0x00007fffd55b0000)
        libcrypt.so.2 => not found
        libpthread.so.0 => /lib/x86_64-linux-gnu/libpthread.so.0 (0x00007f03d15ed000)
        libc.so.6 => /lib/x86_64-linux-gnu/libc.so.6 (0x00007f03d13f0000)
        /lib64/ld-linux-x86-64.so.2 (0x00007f03d1636000)
_sqlite3.cpython-311-x86_64-linux-gnu_failed.so:
warning: debug option `1' unknown; try LD_DEBUG=help
warning: debug option `1' unknown; try LD_DEBUG=help
        linux-vdso.so.1 (0x00007fffccde8000)
        libsqlite3.so.0 => /usr/lib/x86_64-linux-gnu/libsqlite3.so.0 (0x00007f9447eb0000)
        libc.so.6 => /lib/x86_64-linux-gnu/libc.so.6 (0x00007f9447cb0000)
        libm.so.6 => /lib/x86_64-linux-gnu/libm.so.6 (0x00007f9447b61000)
        libpthread.so.0 => /lib/x86_64-linux-gnu/libpthread.so.0 (0x00007f9447b3e000)
        libdl.so.2 => /lib/x86_64-linux-gnu/libdl.so.2 (0x00007f9447b30000)
        /lib64/ld-linux-x86-64.so.2 (0x00007f9448021000)
```

**Comment by pcroland** [comment]
I upgraded to Ubuntu 22.04 and it compiled after.

**Comment by tkundrat** [comment]
I also get the following warning:
```
*** WARNING: renaming "_sqlite3" since importing it failed: /tmp/python-build.20231009215316.14443/Python-3.11.6/build/lib.linux-aarch64-3.11/_sqlite3.c
python-311-arm-linux-gnueabihf.so: undefined symbol: sqlite3_deserialize
```
Tried everything already mentioned in this thread (e.g. updating and installing all mentioned packages), but nothing worked. I've also tried manually `make`-ing python 3.11 from source, but I always get this error.
Does someone have an idea? I'm running Raspbian bullseye aarch64 on an RasPi 4B.

**Comment by tkundrat** [comment]
@pcroland maybe you (or someone else) can open this issue again?

**Comment by native-api** [comment]
Sure. But without a means to reproduce the issue, we can only advise how to diagnose it. Namely, the obvious: check that it successfully links against everything and whether your local `libsqlite` actually has that `sqlite3_deserialize` symbol for the right architecture.

**Comment by native-api** [comment]
> and whether your local `libsqlite` actually has that `sqlite3_deserialize` symbol for the right architecture.

In particular, "`arm-linux-gnueabihf`" looks suspicious. For Aarch64, it normally reads `arm64` or `aarch64`. Make sure that you're building for the right architecture (e.g. that `arch` output is correct and you don't have envvars set that force an architecture for the compiler).

**Comment by yssource** [comment]
OS: Manjaro 23.1.0 Vulcan
Kernel: x86_64 Linux 6.2.16-2-MANJARO

```
➜ ls -al /usr/lib/libsqlite3.so
lrwxrwxrwx 1.4M root 11 Oct 03:42 /usr/lib/libsqlite3.so -> libsqlite3.so.0.8.6
```

```
➜ PYTHON_CONFIGURE_OPTS="--enable-loadable-sqlite-extensions" pyenv install 3.11:latest
Installing Python-3.11.6...
Traceback (most recent call last):
  File "<string>", line 1, in <module>
  File "/home/jimmy/.pyenv/versions/3.11.6/lib/python3.11/sqlite3/__init__.py", line 57, in <module>
    from sqlite3.dbapi2 import *
  File "/home/jimmy/.pyenv/versions/3.11.6/lib/python3.11/sqlite3/dbapi2.py", line 27, in <module>
    from _sqlite3 import *
ModuleNotFoundError: No module named '_sqlite3'
WARNING: The Python sqlite3 extension was not compiled. Missing the SQLite3 lib?
Installed Python-3.11.6 to /home/jimmy/.pyenv/versions/3.11.6
```

**Comment by yssource** [comment]
I fixed my problem, since there is a outdated header and library dependent files under my `/usr/local/include/` `/usr/local/lib/`. It seems python source looking for sqlite3 under `/usr/local/lib` before the system `/usr/lib`.

After removing these files below, Now it would compile python3.11, or 3.12 successfully.
```
/usr/local/include/sqlite3.h
/usr/local/include/sqlite3ext.h
/usr/local/lib/libsqlite3.a
/usr/local/lib/libsqlite3.so
/usr/local/lib/libsqlite3-1.so
```

**Comment by native-api** [comment]
> It seems python source looking for sqlite3 under `/usr/local/lib` before the system `/usr/lib`.

It does. `/usr/local/` is specifically designed to override distro-provided software (see https://en.wikipedia.org/wiki/Filesystem_Hierarchy_Standard).

**Comment by maitreyakv** [comment]
> After removing these files below, Now it would compile python3.11, or 3.12 successfully.

This also helped me fix a similar issue, so I'll describe it in case others are having similar problems. 

I was trying to compile Python against a custom build of sqlite3 on Ubuntu 18.04. I guess in the process of building and installing sqlite3 on my own, I must have partially installed it to `/usr/local/`, in some broken state. I gave up on that whole effort, but every time I tried reinstalling Python 3.11.4 with `pyenv` I was getting `ModuleNotFoundError: No module named '_sqlite3'` during the install. I tried `sudo apt install sqlite3 libsqlite3` like many others online suggested in order to get a fresh and proper install of sqlite3, but I kept getting the error. As @native-api points out, `/usr/local/` overrides distro-provided software (the apt installed version of sqlite3), so my Python builds were constantly finding my broken custom sqlite3 install in `/usr/local/` before the apt-installed version. By removing all those `sqlite.*` and `libsqlite3.*` files I was able to solve the issue, and `pyenv` finally found the apt-installed versions of sqlite3. Thanks @yssource and @native-api, ya'll really saved me a lot of headache.

**Comment by ghsanti** [comment]
For me using conda this was solved using:

```bash
conda remove python
conda install python -c defaults
```

I think otherwise it was installing from conda forge, and even though the defaults are still packaged by conda forge, now `import  sqlite3` loads fine, and even though the file I have in my conda environment is one and the same name (but it must be different within.):

```bash
cd $CONDA_PREFIX
find . -iname "*_sqlite*"
# ./lib/python3.12/lib-dynload/_sqlite3.cpython-312-x86_64-linux-gnu.so
```

**Comment by lanSeFangZhou** [comment]
Does this question fixed？

**Comment by djunsheep** [comment]
met the same problem. dig the whole internet upsidedown but nothing work. BUT! finally I copy a sqlite3 from the original base env, to the one giving errors, an it worked! WTF?

**Comment by djunsheep** [comment]
> Does this question fixed？

I met similar problem and fixed by removing sqlite from the conda env, update every thing, and reinstall the sqlite

**Comment by Ritesh313** [comment]
TLDR:
Just uninstalling the sqlite3 **within the conda environment** fixed it for me: ```conda uninstall sqlite libsqlite```. This removed, updated, downgraded, and installed some packages one of which was GraalPy. It did finish running but threw an import error at the end: ```ImportError: cannot import name '_psutil_linux'```. I checked all the available different sqlite3 files before and after. There was an additional GraalPy version (```/lib/graalpy23.0/modules/_sqlite3.py```) after running the uninstall command.
Regardless, after this I didn't need to install sqlite3 again and everything ran fine. 

More detail:
I had a similar problem with python 3.11 conda environment, I created a new environment with python 3.12 and the issue was still there. As other have pointed it out, it's due to different versions of sqlite3. Specifically I was getting this when trying to use the conda environment as a jupyter notebook kernel
```
Failed to start the Kernel. 
ImportError: /home/riteshchowdhry/miniconda3/envs/py312/lib/python3.12/lib-dynload/_sqlite3.cpython-312-x86_64-linux-gnu.so: undefined symbol: sqlite3_deserialize. 
```
This also can be reproduced by running python in the terminal and trying to import sqlite3:
```
$ python
Python 3.12.2 | packaged by conda-forge | (main, Feb 16 2024, 20:50:58) [GCC 12.3.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import sqlite3
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/riteshchowdhry/miniconda3/envs/py312/lib/python3.12/sqlite3/init.py", line 57, in <module>
    from sqlite3.dbapi2 import *
  File "/home/riteshchowdhry/miniconda3/envs/py312/lib/python3.12/sqlite3/dbapi2.py", line 27, in <module>
    from _sqlite3 import *
ImportError: /home/riteshchowdhry/miniconda3/envs/py312/lib/python3.12/lib-dynload/_sqlite3.cpython-312-x86_64-linux-gnu.so: undefined symbol: sqlite3_deserialize
```
I checked the different available versions of sqlite3 before running the uninstall command:
```
$ find . -iname "*_sqlite*"
./miniconda3/lib/python3.12/lib-dynload/_sqlite3.cpython-312-x86_64-linux-gnu.so
./miniconda3/envs/py312/lib/python3.12/lib-dynload/_sqlite3.cpython-312-x86_64-linux-gnu.so
./miniconda3/pkgs/python-3.11.9-hb806964_0_cpython/lib/python3.11/lib-dynload/_sqlite3.cpython-311-x86_64-linux-gnu.so
./miniconda3/pkgs/sqlalchemy-2.0.34-py311h00e1ef3_0/info/test/test/dialect/test_sqlite.py
./miniconda3/pkgs/python-3.12.4-h5148396_1/lib/python3.12/lib-dynload/_sqlite3.cpython-312-x86_64-linux-gnu.so
./miniconda3/pkgs/python-3.12.2-hab00c5b_0_cpython/lib/python3.12/lib-dynload/_sqlite3.cpython-312-x86_64-linux-gnu.so
./miniconda3/pkgs/python-3.12.7-h5148396_0/lib/python3.12/lib-dynload/_sqlite3.cpython-312-x86_64-linux-gnu.so
```
And then after running it, all of the above were there along with a new one:
```
./miniconda3/pkgs/graalpy-23.0.0-0_graalvm_native/lib/jvm/languages/python/lib/graalpy23.0/modules/_sqlite3.py
```
Not sure if having GraalPy fixed my issue, I haven't tried removing it yet.

**Comment by pierrel-murmuration** [comment]
Hi Stackoverflow, 
I had this bug when building and running a jupyter notebook docker image added with many data science packages.
It appeared from 27 September 2024 while I changed nothing on my side. 
Solution : I switched from python-3.11.6 to python-3.11.10 and also updated conda before the install : 

`FROM jupyter/base-notebook:python-3.11` ==> `FROM quay.io/jupyter/base-notebook:python-3.11.10`

Added: 
`RUN conda update -n base conda -y && \
    mamba update -n base mamba -y && \`
